### PR TITLE
Group 17

### PR DESCRIPTION
**Pull request by sitarakb**
Please read **source/Interfaces/GUI/README.md** in order to compile & run our files/tests

### CODE REVIEW

File: source/Interfaces/GUI/Menu.cpp [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+    case enter:
+      if (!is_empty()) {
+        activate_selected();
+      }
+      break;
+    default:
+        // if input is unknown
+        break;
+  }
+}
```
Comment (Plutolinus) [inline source/Interfaces/GUI/Menu.cpp]: Avoids "magic numbers" & Intuitive interface: The handle_input(int input_code) function takes an int and maps 1, 2, and 3 to directional and enter keys. This interface relies on "magic numbers" and requires the caller to know exactly which integer corresponds to which action (e.g., calling handle_input(1);), making it easy to misuse. It would be much better to define an enum class HandleInput { Up, Down, Enter }; in the header and use that enum as the parameter type. This makes the interface self-documenting and prevents arbitrary integers from being passed.

File: source/Interfaces/GUI/Menu.hpp [hunk 2]
```diff
@@ -0,0 +1,86 @@
...
+
+public:
+
+  Menu();
+
+  /**
+    * add new option to menu
+    * label The display text for option
+    * callback Function to call when option is chosen
+ */
```
Comment (Plutolinus) [inline source/Interfaces/GUI/Menu.hpp]: Informative comments & Consistent style: The comment styles in the header file are inconsistent. The add_option method uses a Doxygen-style block comment /** ... */ (though it's missing the @param tags), but starting from line 40, all subsequent methods use standard // inline comments. Additionally, there is a minor typo on line 40: "in th menu" should be "in the menu". Please unify the commenting style throughout the header file.

File: source/Interfaces/GUI/Menu.cpp
Comment (Plutolinus) [inline source/Interfaces/GUI/Menu.cpp]: Implementation
Menu.hpp (Line 25) & Menu.cpp (Lines 74, 50, 92, etc.)
// Menu.hpp
   int selected_index; // holds -1 if nothing is chosen
// Menu.cpp (Line 74)
  selected_index = (selected_index + 1) % options.size();

Uses modern C++ constructs & No warnings: You are using int selected_index to track the selection, using -1 to represent "no selection". However, options.size() returns a size_t (an unsigned integer). When performing arithmetic like (selected_index + 1) % options.size(), you are mixing signed and unsigned types, which will likely trigger a signed/unsigned mismatch compiler warning when compiling with -Wall and -Wextra. To fix this and embrace modern C++, consider using std::optional<size_t> selected_index; instead of relying on -1. This allows the index type to match size_t consistently.

File: source/Interfaces/GUI/Menu.cpp [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        selected_index = options.size() -1;
+      }
+
+      return true;
+    }
+
+  }
+  return false;
+
+}
```
Comment (Plutolinus) [inline source/Interfaces/GUI/Menu.cpp]: Uses standard library algorithms: The remove_option method uses a manual for loop to find and erase an option. Per the review rubric on avoiding extraneous loops, you should leverage standard library algorithms. In C++23, you can use std::erase_if combined with ranges, or standard algorithms to make this more concise and expressive

File: source/Interfaces/GUI/ImageGrid.cpp
Comment (aryaplayzz) [inline source/Interfaces/GUI/ImageGrid.cpp]: Hi!

So I read through your algorithm, and everything looks pretty solid.

Unfortunately, I tried running your test files and ran into linker problems. I used AI to help troubleshoot it, and after about 7 back-and-forth iterations, it still seems like it does not work on my Mac (Apple Silicon).

I think I may have to wait until we figure out an overall compilation tool (like CMake) before I can properly review your test cases, because I am not entirely sure what else to try, especially since I do not have another machine (like Windows) to test your algorithm on.

I did spend a good amount of time reading your code though, and I really do like the way you implemented it.

One suggestion: I think the documentation could be a little better.

For example, if you could write a quick overview of what the file does and why it is needed, that would be really helpful for any new developer reading your code.

I know this might sound a little weird, but as a new developer looking at code I did not write, it can take a while to read through everything and fully understand it. I personally would rather read comments first than jump straight into the code to understand what the file does (obviously you should read both the comments and the code), but yeah, having a good description at the top of the files would be really nice for new devs reading new code.

File: tests/Interfaces/GUI/ImageManagerTest.cpp
Comment (AB123456kkkkk) [inline tests/Interfaces/GUI/ImageManagerTest.cpp]: For testing and error checking, all of the public member functions including constructors have unit tests.

File: source/Interfaces/GUI/ImageGrid.cpp [hunk 4]
```diff
@@ -0,0 +1,102 @@
+/**
+ * @file ImageGrid.cpp
+ * @author Deni Tepic
+ * 
+ */
+
+#include "ImageGrid.hpp"
+#include "ImageManager.hpp"
+
+using namespace cse498;
```
Comment (hoangtu0701) [inline source/Interfaces/GUI/ImageGrid.cpp]: Using a blanket "using namespace" is generally discouraged I guess. Wrapping the implementations in namespace cse498 { } would be safer and more consistent.

File: source/Interfaces/GUI/ImageGrid.cpp [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+ */
+void ImageGrid::Clear() {
+  for (auto& cell : mCells) {
+    cell.clear();
+  }
+}
+
+/**
+ * resize grid while preserving overlapping region.
+ * *AIGENERATED double for-loop*
```
Comment (hoangtu0701) [inline source/Interfaces/GUI/ImageGrid.cpp]: Might want to clean up the AI attribution tags - "AIGENERATED" in a comment looks a bit rough. A simple "adapted from AI suggestion" or removing it entirely would be cleaner lol

File: source/Interfaces/GUI/Menu.hpp [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+  // check if menu has options
+  // true if menu empty
+  bool is_empty() const;
+
+  // clear all options from the menu
+  void clear();
+
+  // handles keyboard/mouse selection
+  // input_code: int representing user input (1 = up, 2 = down, 3 = enter)
+  void handle_input(int input_code);
```
Comment (JordanSebagh) [inline source/Interfaces/GUI/Menu.hpp]: Even though you define local constants, callers can still pass arbitrary ints and it’s easy to mix up the meaning

File: source/Interfaces/GUI/Menu.cpp
Comment (JordanSebagh) [inline source/Interfaces/GUI/Menu.cpp]: The biggest warning risk is mixing int (selected_index) and size_t (options.size(), indices). Some expressions could trigger signed/unsigned warnings and possibly bugs if selected_index is negative.

File: source/Interfaces/GUI/Text.cpp [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+    } catch (const std::runtime_error& e) {
+        // If default font fails, we'll try to load it again later
+    }
+}
+
+/**
+ * Destructor
+ */
+Text::~Text() {
+    if (mFont) {
```
Comment (gudimaya) [inline source/Interfaces/GUI/Text.cpp]: In this line should the destructor null out the pointers? Like this?

if (mFont) { TTF_CloseFont(mFont); mFont = nullptr; }

**Review by gudimaya — COMMENTED** [review]
I apologize for the late review, I believe your code is really great and I made some suggestions that really are optional that could make the code a little better.

**Review by motowsk2 — COMMENTED** [review]
Submitting comments for review, overall great job! Feel free to reach back out to me if you have any questions on the comments I left.

File: source/Interfaces/GUI/ErrorManager.hpp
Comment (LimEddie19) [inline source/Interfaces/GUI/ErrorManager.hpp]: Great code overall. It is easy to understand and great choice in representing some datatypes. Like the struct for the color choices. The variable names and function names have consistent style (camelCase and PascalCase) and they looked evenly spaced (clang-format)

File: source/Interfaces/GUI/ImageManager.hpp
Comment (Mehver) [inline source/Interfaces/GUI/ImageManager.hpp]: *Some wording in this review was refined with the help of LLM, and some suggestions were identified by the LLM and then reviewed, filtered, and supplemented by me.*

# 1 Class Design

## 1.1 Informative identifiers

Are names of variables, functions, types, etc, clear and helpful for a reader to understand the code. For example, if you need to track the number of students in a class, good variable names might be `num_students` or `student_count`, whereas you don’t want something too short to be confusing (like `N` or `num_s`) or too verbose that it’s annoying (like `total_number_of_students_in_class`).

> [!TIP]
>
> All variable semantics are clear. Member variables, the constructor, and functional methods currently use different naming styles. We will likely receive a company-level unified style guideline later, so please follow it when it becomes available.

## 1.2 Avoids “magic numbers”

Similar to above, you should use constant variables with informative names instead of arbitrary numbers throughout your code, and instead set constants. For example, I have a virtual CPU class with 6 virtual registers, but rather than using “6” throughout, I set `static constexpr size_t NUM_REGS = 6;` at the top of the class to make it clearer what the 6 refers to (and, in this case so that I can change it easily if needed.) Note that if the meaning of the numbers is obvious from context, it may not be an issue, such as initializing a reference count to 1 in a constructor, or an accumulator to zero.

>[!TIP]
>
>Very good!

## 1.3 Intuitive and functional interface

Does the class expose only the necessary functionality, and is it easy to use correctly while hard to misuse? Are member variables private? Is there an obvious function to use for common tasks, with relevant default arguments for parameters? Are any of the functions misleading, like potentially confusing uses of operator overloading? (Feel free to provide suggestions on how to improve the interface for better usability)

> [!NOTE]
>
> My only concern is that `get_texture()` exposes the raw `SDL_Texture*`, which means external code could potentially destroy this resource after obtaining it. It is recommended to at least clarify the ownership in the documentation, or enforce stricter constraints.
>
> https://github.com/CSE498/Spring2026-CompanyA/blob/b18a554d7c48e3286b64406bf8d1f88bbf84778e/source/Interfaces/GUI/ImageManager.hpp#L42
>
> https://github.com/CSE498/Spring2026-CompanyA/blob/b18a554d7c48e3286b64406bf8d1f88bbf84778e/source/Interfaces/GUI/ImageManager.cpp#L92-L107

## 1.4 Informative comments

Do comments help someone new to the code understand it more rapidly? Did the coder properly avoid redundant or obvious comments that clutter the code (e.g., `++x; // increments x`). In general, comments should focus on *why* code is helpful, rather than belabor *what* that code does (though with complex code, the latter is useful too.)

> [!TIP]
>
> Very good!

## 1.5 Consistent style

Are style choices (like naming schemes, brace placement, etc) used in a consistent and informative manner throughout the code? Is there consistency in how different naming conventions are used? (e.g., `camelCase`, `snake_case`, `ALL_CAPS`, or `PascalCase`, or prefixes and suffixes, like `m_` at the beginning of a member variable.).

> [!TIP]
>
> Very good, just ref to 1.1's tip.

## 1.6 Visual appearance

Is the code easy to scan through and visually well structured with good use of whitespace, indentation, and line breaks? Do there appear to be consistent limits on line widths?

> [!NOTE]
>
> For namespace indentation, the style should be clarified soon in the company-wide clang-format configuration. In most cases, not indenting improves readability.

# 2 Implementation

## 2.1 No warnings during compilation

Does the code compile with the `-Wall` and `-Wextra` flags without producing warnings? One option is to add `-Werror` to your list of options, which turns all warnings into errors and thus halts compilation when one occurs.

> [!TIP]
>
> No cli output (including warning) with:
>
> ```bash
> g++ -std=c++23 -Wall -Wextra -Wpedantic -I/workspaces/Spring2026-CompanyA -I/workspaces/Spring2026-CompanyA/third-party/SDL2/include -c ImageManager.cpp
> clang++ -std=c++23 -Wall -Wextra -Wpedantic -I/workspaces/Spring2026-CompanyA -I/workspaces/Spring2026-CompanyA/third-party/SDL2/include -c ImageManager.cpp
> ```

## 2.2 Proper uses of namespaces

Is all code within `namespace cse498 { ... }` ? Does the code avoid the generic `using namespace std` as well as avoiding any namespace pollution in header files that will spread to users of those files.

> [!IMPORTANT]
>
> I noticed that `using namespace cse498;` is used in the `ImageManager.cpp`. I‘d still recommend using `namespace cse498 {}` instead of `using` to avoid potential risks.
>
> https://github.com/CSE498/Spring2026-CompanyA/blob/b18a554d7c48e3286b64406bf8d1f88bbf84778e/source/Interfaces/GUI/ImageManager.cpp#L15

## 2.3 Adheres to the C++ standard

Does the code avoid undefined behavior? If possible, make sure the code compiles and runs properly on different compilers and different platforms.

> [!NOTE]
>
> Using only `assert` for validity checks here may not be sufficiently safe. It is recommended to strengthen the error handling.
>
> https://github.com/CSE498/Spring2026-CompanyA/blob/b18a554d7c48e3286b64406bf8d1f88bbf84778e/source/Interfaces/GUI/ImageManager.cpp#L17-L26

## 2.4 Appropriate Modularity

Are common procedures grouped into useful helper functions? Are helper classes used where needed? Is unnecessary code duplication avoided?

> [!TIP]
>
> Very good, almost no noticeable duplication.

## 2.5 Uses modern C++ constructs

Where appropriate, does the code use range-based for loops, `std::optional`, `std::expected`, or other modern C++ techniques? (In the future we will also look for proper use of templates, move semantics, lambdas, and `constexpr`, but only once we cover those topics).

> [!TIP]
>
> Very good.

## 2.6 Uses standard library algorithms

Does the code avoid extraneous loops when equivalent functionality exists in the standard library? For example, `std::fill` will fill a vector, `std::remove_if` will remove elements from a container based on a condition, or `std::is_sorted` will tell you if the values in a container are in order. See https://en.cppreference.com/w/cpp/algorithm for a full list (it’s LONG, but well worth being familiar with).

> [!TIP]
>
> Very good.

## 2.7 Pointer handling

Is memory management done carefully, with either smart pointers (`std::unique_ptr`, `std::shared_ptr`) or, only if needed, careful use of raw pointers with `new` and `delete`?

> [!NOTE]
>
> You may consider changing this raw pointer to a reference to further clarify ownership. Please use your judgment based on the actual needs.
>
> https://github.com/CSE498/Spring2026-CompanyA/blob/b18a554d7c48e3286b64406bf8d1f88bbf84778e/source/Interfaces/GUI/ImageManager.hpp#L23-L24
>
> https://github.com/CSE498/Spring2026-CompanyA/blob/b18a554d7c48e3286b64406bf8d1f88bbf84778e/source/Interfaces/GUI/ImageManager.cpp#L21-L26

> [!NOTE]
>
> You may consider replacing `SDL_Texture*` with a smart pointer to improve safety.
>
> https://github.com/CSE498/Spring2026-CompanyA/blob/b18a554d7c48e3286b64406bf8d1f88bbf84778e/source/Interfaces/GUI/ImageManager.hpp#L26-L27

> [!IMPORTANT]
>
> Since a custom destructor is already defined / copying has been disabled, it is recommended to follow [[C++ Core Guidelines C.21](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c21-if-you-define-or-delete-any-copy-move-or-destructor-function-define-or-delete-them-all)](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#c21-if-you-define-or-delete-any-copy-move-or-destructor-function-define-or-delete-them-all) and explicitly `default` or `delete` all five special member functions (`dtor` / `copy ctor` / `copy assign` / `move ctor` / `move assign`) to avoid accidental non-movability or unclear semantics caused by implicit generation rules.
>
> If smart pointers are used instead, the necessity of this step is reduced.
>
> https://github.com/CSE498/Spring2026-CompanyA/blob/b18a554d7c48e3286b64406bf8d1f88bbf84778e/source/Interfaces/GUI/ImageManager.hpp#L34-L36

## 2.8 `const` correctness

Are member functions `const` if they don’t change the state of the class? Are parameters (especially reference parameters) marked `const` if they don’t change inside the function?

> [!NOTE]
>
> I noticed that this actually changes the rendering, since the pointer is only shallow `const` while the pointed-to object is modified. Semantically, I would prefer not to mark this function as `const`.
>
> https://github.com/CSE498/Spring2026-CompanyA/blob/b18a554d7c48e3286b64406bf8d1f88bbf84778e/source/Interfaces/GUI/ImageManager.cpp#L109-L138

# 3 Testing and Error Checking

## 3.1 Unit testing

Are unit tests present for all public member functions (including constructors) and well organized? Do they trigger cases that will exercise most private portions of the class’ code?

> [!TIP]
>
> Very good.

## 3.2 Edge cases

Does testing cover most edge cases, as possible? For example, for pointers, does it test null pointers (if they are allowed)? Does it test for empty containers being used? Does it test for 0 or other boundary conditions on parameters?

>[!NOTE]
>
>Here are several suggestions that the LLM and I came up with:
>
>1. `load_image("", path)`: Should an empty name be allowed? (If not, it should throw an exception.)
>2. `load_image(name, "")`: Empty path.
>3. Extreme coordinates in `draw_image` (negative or very large values): SDL usually allows this, but it may be worth adding a smoke test.
>4. Loading different paths with the same name repeatedly: you currently throw on duplicates (which aligns with the “unique name” design), and the tests already cover same-name cases.

## 3.3 Assertions

Are `assert` statements used properly throughout the code to enforce invariants and ensure a valid state? (for example, checking that valid arguments are passed into functions.) Note that asserts should be used only to identify conditions that should never occur in properly working code; if code is expected to recover, error handling must be performed.

>[!NOTE]
>
>Ref to 2.3's note.

## 3.4 Error handling

Are error conditions that can occur during correct execution (such as resource limitations, files not available, or division by zero) checked for and reported back with `std::expected`, other return conditions, or some other technique like exceptions (if your group is comfortable with the speed hit)?

>[!TIP]
>
>Very good.

**Review by Mehver — COMMENTED** [review]
I’ve put all the comments into a single document for your convenience.

File: source/Interfaces/GUI/ErrorManager.hpp [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+    void SetTerminalWarningColor(const std::string& color) { terminalWarningColor = color; }
+    void SetLogErrorColor(const std::string& color)        { logErrorColor = color; }
+    void SetLogWarningColor(const std::string& color)      { logWarningColor = color; }
+
+    /// Prints message and immediately terminates the program.
+    /// Use for unrecoverable states (e.g. failed memory allocation, corrupt game state).
+    void RaiseFatalError(const std::string& message) {
+        PrintInColor(fatalErrorColor, "[FATAL] " + message);
+        std::exit(EXIT_FAILURE);
+    }
```
Comment (LimEddie19) [inline source/Interfaces/GUI/ErrorManager.hpp]: (Including line 28 and 74)

It seems that fatalErrorColor right now is only BOLD. If there was an intentional color designed for it don't forget that. If it was intended to just be BOLD, then you can disregard this comment.

File: source/Interfaces/GUI/Text.cpp [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+
+
+/**
+ * Constructor
+ * @param renderer SDL renderer for drawing
+ */
+Text::Text(SDL_Renderer* renderer)
+    : mContent("")
+    , mFontPath(DEFAULT_FONT)
+    , mColor(255, 255, 255, 255)  // White
```
[the remark this answers is not in the document]
Comment (kianamay) [reply source/Interfaces/GUI/Text.cpp]: good point! thanks

File: source/Interfaces/GUI/ImageManager.hpp [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+
+#ifndef IMAGEMANAGER_HPP
+#define IMAGEMANAGER_HPP
+
+#include <SDL2/SDL.h>
+#include <SDL2/SDL_image.h>
+
+#include <string>
+#include <unordered_map>
+
```
Comment (karensuzue) [inline source/Interfaces/GUI/ImageManager.hpp]: You should add `#include <memory>` here to avoid compilation errors.

File: source/Interfaces/GUI/ImageGrid.cpp [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+#include "ImageGrid.hpp"
+#include "ImageManager.hpp"
+
+using namespace cse498;
+
+/**
+ * construct a grid with dimensions and tile size.
+ */
+ImageGrid::ImageGrid(size_t width, size_t height,
+                     int tile_width, int tile_height)
```
Comment (karensuzue) [inline source/Interfaces/GUI/ImageGrid.cpp]: This does not match the constructor signature in `ImageGrid.hpp`, which is `ImageGrid(size_t, size_t, size_t, size_t)`

Comment (karensuzue) [reply source/Interfaces/GUI/ImageGrid.cpp]: -R -C

File: source/Interfaces/GUI/ImageGrid.cpp [hunk 12]
```diff
@@ -0,0 +1,102 @@
...
+  return mCells[ToIndex(x, y)];
+}
+
+/**
+ * set image name at specified cell.
+ */
+void ImageGrid::SetCell(size_t x, size_t y,
+                        const std::string& image_name)
+{
+  assert(IsValid(x, y));
```
Comment (karensuzue) [inline source/Interfaces/GUI/ImageGrid.cpp]: Same issue with using assertions for input validation

File: source/Interfaces/GUI/ImageManager.cpp [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+     * @param name unique name of the image
+     * @param file_path path to the image file
+     */
+    void ImageManager::load_image(const std::string& name,
+                                  const std::string& file_path)
+    {
+        // empty string checks
+        if (name.empty() || file_path.empty())
+        {
+            throw std::invalid_argument("Image name or file path can't be empty.");
```
Comment (karensuzue) [inline source/Interfaces/GUI/ImageManager.cpp]: I know we recently decided as a company not to use exceptions, but I won't be enforcing that here

**Review by nahumj — APPROVED** [review]
See private channel on discord for grade.

File: tests/Interfaces/GUI/ImageManagerTest.cpp [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+    SDLMock mock;
+    ImageManager manager(mock.renderer);
+
+    // load image
+    std::string testImage = "Interfaces/GUI/images/ImageManagerTest.png";
+    manager.load_image("test_img", testImage);
+
+    // confirm it returns true
+    CHECK(manager.draw_image("test_img", 10, 10) == true);
+}
```
[the remark this answers is not in the document]
Comment (sitarakb) [reply tests/Interfaces/GUI/ImageManagerTest.cpp]: Thank you. I have now added a test that uses SDL_RenderReadPixels to check for pixels being displayed.

File: source/Interfaces/GUI/Menu.hpp [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+
+public:
+
+  Menu();
+
+  /**
+    * add new option to menu
+    * label The display text for option
+    * callback Function to call when option is chosen
+ */
```
[the remark this answers is not in the document]
Comment (anaghaj9) [reply source/Interfaces/GUI/Menu.hpp:38]: Sounds good! Thank you!

File: source/Interfaces/GUI/Menu.hpp [hunk 16]
```diff
@@ -0,0 +1,95 @@
...
+
+  // move selection to the previous option
+  void select_previous();
+
+  // selects option by it's index
+  // index: index to select
+  void select_option(size_t index);
+
+  // activates selected options, triggers its callback
+  void activate_selected();
```
[the remark this answers is not in the document]
Comment (anaghaj9) [reply source/Interfaces/GUI/Menu.hpp:62]: Yeah, I agree. I didn't think about it like that, but now that you mention that, it makes sense!

File: source/Interfaces/GUI/Menu.hpp [hunk 17]
```diff
@@ -0,0 +1,95 @@
...
+  std::string get_option_label(size_t index) const;
+
+  // check if menu has options
+  // true if menu empty
+  bool is_empty() const noexcept;
+
+  // clear all options from the menu
+  void clear();
+
+  enum class InputCode
```
[the remark this answers is not in the document]
Comment (anaghaj9) [reply source/Interfaces/GUI/Menu.hpp:77]: Yeah, I try to make all my function names snake_case and all my variables PascalCase for readability purposes!

File: source/Interfaces/GUI/Menu.hpp [hunk 18]
```diff
@@ -0,0 +1,95 @@
...
+      up = 1,
+      down = 2,
+      enter = 3
+  };
+  // handles keyboard/mouse selection
+  // input_code: int representing user input (1 = up, 2 = down, 3 = enter)
+  void handle_input(InputCode input_code);
+
+  // draw function for menu button
+  void draw(SDL_Renderer* renderer, int x, int y, int width, int height);
```
[the remark this answers is not in the document]
Comment (anaghaj9) [reply source/Interfaces/GUI/Menu.hpp:88]: Yeah, agree! I'll chat with my group about that.

File: source/Interfaces/GUI/Menu.cpp [hunk 19]
```diff
@@ -0,0 +1,250 @@
...
+
+  if (!callback) {
+    throw std::invalid_argument("Menu option callback cannot be null");
+  }
+
+  MenuOption option;
+  option.label = label;
+  option.callback = callback;
+
+  options.push_back(option);
```
[the remark this answers is not in the document]
Comment (anaghaj9) [reply source/Interfaces/GUI/Menu.cpp:41]: Ah good to know that! This was helpful as I go through and make changes to some of my error handling functions and have a similar push_back call.

File: source/Interfaces/GUI/Menu.cpp [hunk 20]
```diff
@@ -0,0 +1,250 @@
...
+bool Menu::remove_option(const std::string& label) {
+  for(size_t i = 0; i < options.size(); i++){
+    if (options[i].label == label) {
+      options.erase(options.begin() + static_cast<int>(i));
+
+      if (options.empty()) {
+        selected_index = std::nullopt;
+      }
+
+      else if (selected_index.has_value() && *selected_index >= options.size()) {  // ADD *
```
[the remark this answers is not in the document]
Comment (anaghaj9) [reply source/Interfaces/GUI/Menu.cpp:63]: Thanks!

File: source/Interfaces/GUI/Menu.cpp [hunk 21]
```diff
@@ -0,0 +1,250 @@
...
+  if (!selected_index.has_value()) {
+    selected_index = options.size() - 1;
+  }
+
+  else if (*selected_index == 0) {  // ADD *
+    selected_index = options.size() - 1;
+  }
+
+  else {
+    selected_index = *selected_index - 1;  // ADD *
```
[the remark this answers is not in the document]
Comment (anaghaj9) [reply source/Interfaces/GUI/Menu.cpp:121]: Thank you!

File: source/Interfaces/GUI/Menu.cpp [hunk 22]
```diff
@@ -0,0 +1,250 @@
...
+/**
+ * Sets selections to the option at the specified index
+ * @param index 0-based position in the menu
+ */
+void Menu::select_option(size_t index) {
+  if (index >= options.size()) {
+    throw std::out_of_range("Menu option index out of range");
+  }
+
+  selected_index = static_cast<int>(index);
```
[the remark this answers is not in the document]
Comment (anaghaj9) [reply source/Interfaces/GUI/Menu.cpp:135]: Ah, yes it is. Not sure what I was thinking here haha, thanks for pointing that out!

File: source/Interfaces/GUI/Menu.cpp [hunk 23]
```diff
@@ -0,0 +1,250 @@
...
+void Menu::clear() {
+  options.clear();
+  selected_index = std::nullopt;
+}
+
+/**
+ * Handles user inputted keyboard option
+ * @param input_code values for keyboard functions
+ */
+void Menu::handle_input(InputCode input_code) {
```
[the remark this answers is not in the document]
Comment (anaghaj9) [reply source/Interfaces/GUI/Menu.cpp:180]: Good idea! I didn't think about testing terminal-based menus.

File: source/Interfaces/GUI/Menu.cpp [hunk 24]
```diff
@@ -0,0 +1,250 @@
...
+  }
+
+  // Return silently if no renderer provided
+  if (!renderer) {
+    return;
+  }
+
+  // Create Text object for rendering
+  Text menuText(renderer);
+  menuText.SetSize(18);  // Font size for menu options
```
[the remark this answers is not in the document]
Comment (anaghaj9) [reply source/Interfaces/GUI/Menu.cpp:214]: Ah yes! I made a struct with all the values

File: tests/Interfaces/GUI/MenuTest.cpp [hunk 25]
```diff
@@ -0,0 +1,395 @@
...
+  menu.select_option(1);
+  CHECK(menu.get_selected_index().value() == 1);
+
+  menu.select_option(2);
+  CHECK(menu.get_selected_index().value() == 2);
+
+  menu.select_option(0);
+  CHECK(menu.get_selected_index().value() == 0);
+}
+
```
[the remark this answers is not in the document]
Comment (anaghaj9) [reply tests/Interfaces/GUI/MenuTest.cpp:105]: Ooo yeah, good point! I like that idea to test how the code functions when I move between selection methods.

File: tests/Interfaces/GUI/MenuTest.cpp [hunk 26]
```diff
@@ -0,0 +1,395 @@
...
+TEST_CASE("Handle input - navigation", "[Menu]")
+{
+  Menu menu;
+  menu.add_option("First", [](){});
+  menu.add_option("Second", [](){});
+
+  // Simulate down arrow (code 2)
+  menu.handle_input(cse498::Menu::InputCode::down);
```
[the remark this answers is not in the document]
Comment (anaghaj9) [reply tests/Interfaces/GUI/MenuTest.cpp:176]: Good to know, I implemented a test for the SDL input handling

File: source/Interfaces/GUI/Text.cpp [hunk 27]
```diff
@@ -0,0 +1,274 @@
...
+        }
+        ttf_initialized = true;
+    }
+
+    // Load default font
+    try {
+        ReloadFont();
+    } catch (const std::runtime_error& e) {
+        // If default font fails, we'll try to load it again later
+    }
```
[the remark this answers is not in the document]
Comment (kianamay) [reply source/Interfaces/GUI/Text.cpp:47]: added a warning to cerr

File: source/Interfaces/GUI/Text.cpp (hunk 7 above)
Comment (kianamay) [reply source/Interfaces/GUI/Text.cpp:54]: mFont = nullptr; 

added this as a fallback

File: source/Interfaces/GUI/Text.cpp [hunk 28]
```diff
@@ -0,0 +1,274 @@
...
+}
+
+/**
+ * Move assignment operator
+ */
+Text& Text::operator=(Text&& other) noexcept {
+    if (this != &other) {
+        if (mFont) {
+            TTF_CloseFont(mFont);
+        }
```
[the remark this answers is not in the document]
Comment (kianamay) [reply source/Interfaces/GUI/Text.cpp:84]: I was thinking that the move constructor doesn't need to null mFont after construction because the object is being newly created but the move assignment needs it because it's replacing an existing object so i put it here instead

File: source/Interfaces/GUI/Text.cpp [hunk 29]
```diff
@@ -0,0 +1,274 @@
...
+void Text::SetColor(unsigned char r, unsigned char g, unsigned char b, unsigned char a) {
+    mColor.r = r;
+    mColor.g = g;
+    mColor.b = b;
+    mColor.a = a;
+}
+
+void Text::SetFont(const std::string& font_name) {
+    mFontPath = font_name;
+    ReloadFont();  // May throw if font doesn't exist
```
[the remark this answers is not in the document]
Comment (kianamay) [reply source/Interfaces/GUI/Text.cpp:118]: saves and uses previous font

File: source/Interfaces/GUI/Text.cpp [hunk 30]
```diff
@@ -0,0 +1,274 @@
...
+    , mColor(DEFAULT_COLOR)  // White
+    , mFontSize(DEFAULT_SIZE)
+    , mBold(false)
+    , mItalic(false)
+    , mRenderer(renderer)
+    , mFont(nullptr)
+{
+    // Initialize SDL_ttf if not already initialized
+    static bool ttf_initialized = false;
+    if (!ttf_initialized) {
```
[the remark this answers is not in the document]
Comment (kianamay) [reply source/Interfaces/GUI/Text.cpp:35]: got it, added more documentation

File: source/Interfaces/GUI/Text.cpp [hunk 31]
```diff
@@ -0,0 +1,274 @@
...
+    mColor.a = a;
+}
+
+void Text::SetFont(const std::string& font_name) {
+    mFontPath = font_name;
+    ReloadFont();  // May throw if font doesn't exist
+}
+
+void Text::SetSize(int point_size) {
+    assert(point_size > 0 && "Font size must be positive");
```
[the remark this answers is not in the document]
Comment (kianamay) [reply source/Interfaces/GUI/Text.cpp:122]: thanks, i changed this

File: source/Interfaces/GUI/Text.cpp [hunk 32]
```diff
@@ -0,0 +1,274 @@
...
+
+    if (point_size <= 0) {
+        return;  // Assert should catch this in debug mode
+    }
+
+    mFontSize = point_size;
+    ReloadFont();  // Reload font with new size
```
[the remark this answers is not in the document]
Comment (kianamay) [reply source/Interfaces/GUI/Text.cpp:129]: i added some safeguards to this

File: source/Interfaces/GUI/Text.cpp [hunk 33]
```diff
@@ -0,0 +1,274 @@
...
+    mFontSize = DEFAULT_SIZE;
+    mColor = DEFAULT_COLOR;   // White
+    mBold = false;
+    mItalic = false;
+
+    ReloadFont();
+}
+
+void Text::Draw(int x, int y) {
+    // User error: empty content - silently return
```
[the remark this answers is not in the document]
Comment (kianamay) [reply source/Interfaces/GUI/Text.cpp:157]: got it, fixed

File: source/Interfaces/GUI/Text.cpp [hunk 34]
```diff
@@ -0,0 +1,274 @@
...
+}
+
+void Text::SetRenderer(SDL_Renderer* r) {
+    mRenderer = r;
+}
+
+void Text::ReloadFont() {
+    // Close existing font if any
+    if (mFont) {
+        TTF_CloseFont(mFont);
```
[the remark this answers is not in the document]
Comment (kianamay) [reply source/Interfaces/GUI/Text.cpp:238]: Changed to unique ptr, thank you

File: source/Interfaces/GUI/Text.hpp [hunk 35]
```diff
@@ -0,0 +1,223 @@
...
+protected:
+    std::string mContent;           ///Text content to render
+    std::string mFontPath;          ///Path to current font file
+    Color mColor;                   ///Text color
+    int mFontSize;                  ///Font size in points
+    bool mBold;                     ///Bold style flag
+    bool mItalic;                   ///Italic style flag
+
+    SDL_Renderer* mRenderer;        ///SDL renderer for drawing
+    TTF_Font* mFont;                ///Loaded SDL font object
```
[the remark this answers is not in the document]
Comment (kianamay) [reply source/Interfaces/GUI/Text.hpp:53]: changed to unique ptr

File: source/Interfaces/GUI/Text.hpp [hunk 36]
```diff
@@ -0,0 +1,223 @@
...
+
+    // Disable copying to avoid issues with font management
+    Text(const Text&) = delete;
+    Text& operator=(const Text&) = delete;
+
+    // Allow moving
+    Text(Text&& other) noexcept;
+    Text& operator=(Text&& other) noexcept;
+
+     /// Dimensions structure for text measurements
```
[the remark this answers is not in the document]
Comment (kianamay) [reply source/Interfaces/GUI/Text.hpp:88]: yes

File: source/Interfaces/GUI/Text.hpp [hunk 37]
```diff
@@ -0,0 +1,208 @@
...
+     * @param y Y coordinate (top edge of text)
+     * @note Returns silently if content is empty (user error)
+     */
+    void Draw(int x, int y);
+
+    /**
+     * Get the width of the rendered text in pixels
+     * @return Width in pixels, or 0 if content is empty
+     */
+    int GetWidth() const;
```
[the remark this answers is not in the document]
Comment (kianamay) [reply source/Interfaces/GUI/Text.hpp:201]: I considered this but ultimatley decided against it to keep the code more flexible
